### PR TITLE
Added Webhooks

### PR DESCRIPTION
**Pull request by UnderMybrella**
### Prerequisites
* [x] I have read and understood the [contribution guidelines](CONTRIBUTING.md)
* [x] This pull request follows the code style of the project
* [x] I have tested this feature thoroughly
  * [Proof of testing](http://imgur.com/a/tWDIB)

### Changes Proposed in this Pull Request
* Adds IWebhook, and the implementing object Webhook
* Adds an ignore function for webhook warnings (in a similar fashion to messages for channels)
* Adds the WEBHOOK Endpoint
* Handle the WEBHOOK_UPDATE message, break it into WebhookUpdateEvent, WebhookCreateEvent, and WebhookDeleteEvent
* Added methods to IChannel and IGuild for creating and getting webhooks. VoiceChannel and PrivateChannel instances through UnsupportedOperationException when called.
* Added WebhookList, as a method of caching and retrieving webhooks.
* Added the other embed objects; does not interfere with attachments or reactions.

### CODE REVIEW

**Comment by ArsenArsen** [comment]
This is a very misleading name. Should be changed into ``Implementing webhooks`` or something in my opinion. Didn't take a too good look at the code just yet.

File: src/main/java/sx/blah/discord/handle/obj/IMessage.java [hunk 1]
```diff
@@ -274,20 +275,97 @@ public String getUrl() {
...
+		 * @return An object containing information about the author for the embedded media. Cna be null.
+		 */
+		public IEmbedded.IEmbedAuthor getAuthor();
+
+		public List<IEmbedded.IEmbedField> getEmbeddedFields();
+
+		/**
+		 * Represents an embedded footer object
+		 */
+		interface IEmbedFooter {
```
Comment (chrislo27) [inline src/main/java/sx/blah/discord/handle/obj/IMessage.java]: Unnecessary public modifier on an interface method

File: src/main/java/sx/blah/discord/util/WebhookList.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	 * This is true if the client object has permission to read this channel's webhooks.
+	 */
+	private volatile boolean hasPermission;
+
+	/**
+	 * @param client  The client for this list to respect.
+	 * @param channel The channel to retrieve webhooks from.
+	 */
+	public WebhookList(IDiscordClient client, IChannel channel) {
+		if (channel instanceof IVoiceChannel)
```
Comment (chrislo27) [inline src/main/java/sx/blah/discord/util/WebhookList.java]: This should also include IPrivateChannel

**Comment by UnderMybrella** [comment]
So remove the WebhooktList class, change it to a List<IWebhook>, and provide something like a loadWebhooks() command?

**Comment by austinv11** [comment]
No, the user should not need to manually update webhooks, it should be updated automatically via the websocket class. However you are right about the webhooks being stored in a CopyOnWriteArrayList.

**Comment by UnderMybrella** [comment]
Alright, that's what I originally had. The issue I ran into though was that Discord only sent the webhook update information when a webhook was actually updated, and so I had to hook into guildCreate and the like to keep an updated list. So would that be preferable to the WebhookList class?

**Comment by austinv11** [comment]
That is correct

**Comment by UnderMybrella** [comment]
So for the changes:
Change the Channel's webhook list to a CopyOnWriteArrayList of IWebhook instances
Hook into guildCreate, channelUpdate, guildRoleUpdate, and guildMemberUpdate, to update the webhook cache

Sorry for so many questions, just want to make sure I get this all right.

**Comment by austinv11** [comment]
Yes, that is correct

**Comment by austinv11** [comment]
Can you redo the PR? It's difficult to tell what you actually changed and there are merge conflicts.

**Comment by UnderMybrella** [comment]
Ah yep, sorry. #159

**Pull request closed without merging**